### PR TITLE
Jenkins host deployment. And puppet roles for web stack.

### PR DESCRIPTION
**Pull request by vecinomio**

### CODE REVIEW

File: ops/cloudformation/jenkins.yml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                done
+            }
+            retryCommand 5 10 'aws s3 cp "s3://${S3Path}/scripts/jenkins/mount_volume-${ver}.sh" - | bash -l -s -- --hostname jenkins'
+            rpm -Uvh https://yum.puppet.com/puppet5-release-el-7.noarch.rpm
+            yum -y update
+            retryCommand 5 10 'yum install -y puppet-agent'
+            export PATH=$PATH:/opt/aws/bin/:/opt/puppetlabs/bin/:/opt/puppetlabs/puppet/bin
+            retryCommand 5 10  'aws s3 cp "s3://${S3Path}/scripts/jenkins/puppet-${pupver}.tar" .'
+            tar -C /etc/puppetlabs/ -xvf puppet-${pupver}.tar
+            puppet apply --test /etc/puppetlabs/code/environments/production/manifests/site.pp && echo "Finished puppet"
```
Comment (Emperorex) [inline ops/cloudformation/jenkins.yml]: I would put this command in retry logic. But pay attention about non standard success exit codes:
`--detailed-exitcodes`
`Provide extra information about the run via exit codes; only works if '--test' or '--onetime' is also specified. If enabled, 'puppet agent' will use the following exit codes:`

`0: The run succeeded with no changes or failures; the system was already in the desired state.`

`1: The run failed, or wasn't attempted due to another run already in progress.`

`2: The run succeeded, and some resources were changed.`

`4: The run succeeded, and some resources failed.`

`6: The run succeeded, and included both changes and failures.`

File: ops/puppet/Puppetfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+forge 'https://forge.puppet.com'
+
+# Modules from the Puppet Forge
```
Comment (Emperorex) [inline ops/puppet/Puppetfile]: I would delete all commented lines if it is not necessary

File: ops/puppet/code/modules/profile/manifests/app.pp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      ensure => directory,
+      owner  => 'tomcat',
+      group  => 'tomcat',
+      mode   => '0755',
+    }
+
+    archive { $filename:
+      path          => "/tmp/${filename}",
+      source        => 'https://devopsa3-simple-java-app.s3.amazonaws.com/java-tomcat-v3.zip',
+#      checksum      => 'f2aaf16f5e421b97513c502c03c117fab6569076',
```
Comment (Emperorex) [inline ops/puppet/code/modules/profile/manifests/app.pp]: I would delete all commented lines if it is not necessary

File: ops/puppet/code/modules/func/manifests/ami_extras.pp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+define func::ami_extras(
+  Enum['present'] $ensure = present,
+  ) {
+  $pkg = $name
+  exec { "amazon-linux-extras install -y $pkg":
+  unless => "amazon-linux-extras list | grep -q '${pkg}=.*enabled'",
+  path   => '/usr/bin',
+  }
+}
```
Comment (Emperorex) [inline ops/puppet/code/modules/func/manifests/ami_extras.pp]: Need to add one line

**Comment by vecinomio** [comment]
The Jenkins host has been added in to the project. 
This Pull Request contains:
template with resources needed to create Persistent EBS for Jenkins-host;
template with resources needed to create Jenkins-host;
script "mount-volume" needed to check and attach Persistent EBS to Jenkins-host;
directory with puppet files, needed to provision Jenkins to the instance.

File: ops/puppet/code/modules/func/manifests/ami_extras.pp (hunk 4 above)
Comment (IYermakov) [reply ops/puppet/code/modules/func/manifests/ami_extras.pp]: > Need to add one line

But it works..?

File: ops/cloudformation/jenkins.yml (hunk 1 above)
Comment (vecinomio) [reply ops/cloudformation/jenkins.yml]: Added retry function for puppet command.

File: ops/cloudformation/jenkins.yml [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                    $COMMAND
+                    if [ $? -eq 2 ]
+                    then
+                      echo "Success"
+                      break
+                    else
+                      echo $?
+                    fi
+                done
+            }
```
Comment (IYermakov) [inline ops/cloudformation/jenkins.yml]: would leave only one function with a suchlike logic:
[[ ! ${COMMAND} =~ ^[0,2,4,6]+$ ]] && break

File: ops/cloudformation/jenkins.yml [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
```
Comment (IYermakov) [inline ops/cloudformation/jenkins.yml]: I can't agree with two almost identical functions.
Let's disable detailed exit codes in puppet or leave the second function with a suchlike logic:
if [ $? -ne 1 ] then break

Also let's get rid of unnecessary echo's

**Pull request merged**